llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -368,12 +368,17 @@ static Instruction *foldShiftOfShiftedBinOp(BinaryOperator &I,
 
   // Find a matching one-use shift by constant. The fold is not valid if the sum
   // of the shift values equals or exceeds bitwidth.
-  // TODO: Remove the one-use check if the other logic operand (Y) is constant.
   Value *X, *Y;
-  auto matchFirstShift = [&](Value *V) {
+  auto matchFirstShift = [&](Value *V, Value *W) {
     APInt Threshold(Ty->getScalarSizeInBits(), Ty->getScalarSizeInBits());
-    return match(V,
-                 m_OneUse(m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0)))) &&
+    bool matchFirst;
+    if (match(W, m_ImmConstant()))
+      matchFirst = match(V, (m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))));
+    else
+      matchFirst =
+          match(V, m_OneUse(m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))));
+
+    return matchFirst &&
```

Comment:
```suggestion
    return match(V, m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))) &&
           (V->hasOneUse() && match(W, m_ImmConstant()))
```
No need to duplicate the match.

Reply:
Not done.

Reply:
Your suggestion changes the behavior of an unrelated test and undoes the transform.